source/MaterialXGenShader/ShaderTranslator.cpp

```diff
@@ -0,0 +1,196 @@
+//
+// TM & (c) 2017 Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved.  See LICENSE.txt for license.
+//
+
+#include <MaterialXCore/Util.h>
+#include <MaterialXGenShader/ShaderTranslator.h>
+#include <iostream>
+
+namespace MaterialX
+{
+
+//
+// ShaderTranslator methods
+//
+
+ShaderTranslator::ShaderTranslator(ConstDocumentPtr doc) :
+    _doc(doc)
+{
+    loadShadingTranslations();
+}
+
+void ShaderTranslator::loadShadingTranslations()
+{
+    for (NodeDefPtr node : _doc->getNodeDefs())
+    {
+        if (node->getNodeGroup() == NodeDef::TRANSLATION_NODE_GROUP)
+        {
+            _translationNodes.insert(node->getNodeString());
+
+            // Parsing translation nodes
+            size_t pos = node->getNodeString().find("_to_");
+            string start = node->getNodeString().substr(0, pos);
+            string end = node->getNodeString().substr(pos + 4);
+            std::unordered_map<string, StringSet>::const_iterator it = _shadingTranslations.find(start);
+
+            if (it != _shadingTranslations.end())
+            {
+                _shadingTranslations[start].insert(end);
+            }
+            else
+            {
+                _shadingTranslations[start] = { end };
+            }
+        }
+    }
+}
+
+void ShaderTranslator::connectToTranslationInputs(ShaderRefPtr shaderRef)
+{
+    for (BindInputPtr bindInput : shaderRef->getBindInputs())
+    {
+
+        OutputPtr output = bindInput->getConnectedOutput();
+        if (output)
+        {
+            NodeGraphPtr translationNg = output->getDocument()->getNodeGraph("NG_" + _translationNode->getCategory());
```

Comment:
For consistency with other MaterialX code, I would name this variable "translationGraph", and similarly name the NodeGraph class member "_graph" instead of "_ng".